### PR TITLE
=coll Add quick return for Iterator drop and take.

### PR DESCRIPTION
**Pull request by He-Pin**
Motivation:
Do quick return when we know it. 

Currently, the `take` and `drop` both leverage the `splitIterator` and the `upperbound` is `-1` sometime, and can not do a quick return. but in the `take` and `drop` function, we knew more information of that the caller want to do.

Result:
Quick return for some eage cases, and more check for others .

BTW, I think the current `take` and `drop` with an `Int` parameter is wrong, what if the underling iterator is a LinkedList?

### CODE REVIEW

File: test/junit/scala/collection/IteratorTest.scala [hunk 1]
```diff
@@ -331,6 +341,17 @@ class IteratorTest {
...
   }
+
+  @Test def takeNegativeReturnsEmpty(): Unit = {
+    val iterator = List(1).iterator
+    assertEquals(iterator.take(Int.MinValue), Iterator.empty)
+  }
+
+  @Test def takeMaxValueReturnsSelf(): Unit = {
+    val iterator = List(1).iterator
+    assertEquals(iterator.take(Int.MaxValue), iterator)
```
Comment (som-snytt) [inline test/junit/scala/collection/IteratorTest.scala]: `assertSame` to assert `eq`

File: test/junit/scala/collection/IteratorTest.scala [hunk 2]
```diff
@@ -213,6 +213,16 @@ class IteratorTest {
...
   }
 
+  @Test def dropNegativeReturnsSelf():Unit= {
+    val iterator = List(1).iterator
+    assertEquals(iterator.drop(Int.MinValue), iterator)
+  }
+
+  @Test def dropMaxValueReturnsEmpty() :Unit = {
+    val iterator = List(1).iterator
+    assertEquals(iterator.drop(Int.MaxValue), Iterator.empty)
```
Comment (som-snytt) [inline test/junit/scala/collection/IteratorTest.scala]: `assertFalse(it.hasNext)`

Comment (He-Pin) [reply test/junit/scala/collection/IteratorTest.scala]: Thanks, as your prevous review comment, I think I need to change it to equal.

**Comment by som-snytt** [comment]
There are a few issues or PRs where people ask about how to handle boundary conditions.

I'd be reluctant to introduce change without a compelling reason and understanding the test regime (collections-laws).

I see my `SliceIterator` from 2.12 was "reintroduced" in strawman, which is why the class comment has the wrong formatting. (!)

"Introduce a bounds check" also introduces overhead, so the change is not free, even if it looks trivial.

I haven't looked at what it does, but I see that there are overrides of `sliceIterator` with semantics. There must be a reason for all that code, right?

**Comment by He-Pin** [comment]
That's all becase the `SliceIterator` use a special `-1` to track to linked/bounded one, and the the `take` and `drop` are both using `SliceIterator` , and yes some duplicated check is done in both here and then in the `SliceIterator` which is not that nice.

I'm not a daily Scala developer, using Java for $Work, so my background may hurt me here, I just checked the OpenJDK's Stream `skip` , where `<0` throws and `==0` just return this, and for `limit`, when `<0` throws too.

**Comment by SethTisue** [comment]
second call for opinions from @scala/collections

**Review by iusildra — APPROVED** [review]
Looks fine to me!

**Comment by lrytz** [comment]
I'm not convinced this is worth it. Why check for `<=0` in `Iterator.drop` but not in `IterableOps`? There are probably houndreds of edge cases in the collections library we could identify earlier, but that would come with a tax on code readability / maintainability.

**Comment by lrytz** [comment]
Closing based on my comment above.
